### ISSUE TITLE
Enable -Xsource:3 

### ISSUE THREAD

**Issue by ghostbuster91** [disneystreaming/smithy4s#1584]
This should be enabled for both scala 2.12 and 2.13 builds. 
It ensures that the code behaves in the same way across all language versions. 

This fixes a bug in scala 2.12 and 2.13 where `copy` method is still accessible even though the primary constructor has been marked as `private`. 

Above is binary incompatible change so I guess this goes into 0.19 series.

**Comment by ghostbuster91** [comment]
I will happily volunteer for this.

**Comment by kubukoz** [comment]
I would suggest checking `-Xsource:3-cross`, apparently it's better suited for libs that are cross-built.

**Comment by Baccata** [comment]
Go for it

**Comment by ghostbuster91** [comment]
As usual, it was easier said than done. 

Turns out scala 2.12 does not support `Xsource:3-cross` atm. I asked if there are any plans to backport it [here](https://github.com/scala/scala/pull/10709#issuecomment-2396514517). 

Without this we cannot use these new features like the improved semantics for copy modifiers, meaning that if we want to be better at preserving binary compatibility we have to stop using case classes in public api. Enabling it only for 2.13 & 3 doesn't make sense imo.

We could enable `Xsource:3` that fails when a code with different semantics is detected but this isn't fully supported by scala 2.12 either as the following compiles without any errors/warnings:

```scala
//> using scala 2.12
//> using options -Xsource:3

case class Person private (age: Int, name: String)
```

**Comment by Baccata** [comment]
>  we have to stop using case classes in public api 

I agree that ideally this should be a goal. We even had started on it, but lay-offs and reprioritisations forced us to park this effort : https://github.com/disneystreaming/smithy4s/pull/1391

**Comment by lewisjkl** [comment]
Seeing as how `Xsource:3-cross` is not going to be supported in Scala 2.12 (per the issue linked above), I'm wondering if we should remove this issue from the 0.19 milestone. Any objections?

**Comment by kubukoz** [comment]
Can we add it in 2.13 at least?

**Comment by lewisjkl** [comment]
> Can we add it in 2.13 at least?

To be honest, I'm not really sure what the implications of that would be. I was just going off of the prior comment in this thread "Enabling it only for 2.13 & 3 doesn't make sense imo."

**Comment by kubukoz** [comment]
I think the main benefit is that we can offer better binary compatibility guarantees on case classes, without a gigantic effort of rewriting those. `case class ... private(...)` is already a significant gain in those terms (both 2.13 and 3).

2.12 already has known cases of being bin incompatible where the newer versions are compatible, so this doesn't strike me as something we should avoid doing for the sake of consistency.

**Comment by lewisjkl** [comment]
> I think the main benefit is that we can offer better binary compatibility guarantees on case classes, without a gigantic effort of rewriting those. `case class ... private(...)` is already a significant gain in those terms (both 2.13 and 3).
> 
> 2.12 already has known cases of being bin incompatible where the newer versions are compatible, so this doesn't strike me as something we should avoid doing for the sake of consistency.

Okay cool, I'm good to go forward with it as well then.

**Comment by ghostbuster91** [comment]
> I think the main benefit is that we can offer better binary compatibility guarantees on case classes, without a gigantic effort of rewriting those. case class ... private(...) is already a significant gain in those terms (both 2.13 and 3).

This means that you will be able to do some changes in scala 2.13 and 3 that are binary compatible with lower efforts but you will break scala 2.12. Just wanted to highlight that so that everyone is aware.

**Comment by kubukoz** [comment]
> but you will break scala 2.12.

Maybe at some point within 0.19 we'll drop it :) in that case, it'd be too late to add `-Xsource:3`, but if we start with it, we may have an easier bincompat future.

Remember that 0.18 has been going on for 2.5 years. Who knows if 2.12 will still be around 2 years after launching 0.19!

**Comment by kubukoz** [comment]
closing as done in #1878, we can't do it for 2.12 so there's no reason to keep this open
